File: rewrite-xml/src/main/java/org/openrewrite/xml/internal/rpc/XmlSender.java
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.xml.internal.rpc;

import org.openrewrite.Tree;
import org.openrewrite.rpc.RpcSendQueue;
import org.openrewrite.xml.XmlVisitor;
import org.openrewrite.xml.tree.Content;
import org.openrewrite.xml.tree.Xml;

import java.util.List;

public class XmlSender extends XmlVisitor<RpcSendQueue> {

    @Override
    public Xml preVisit(Xml x, RpcSendQueue q) {
        q.getAndSend(x, Tree::getId);
        q.getAndSend(x, Xml::getPrefix);
        q.getAndSend(x, Tree::getMarkers);
        return x;
    }

    @Override
    public Xml visitDocument(Xml.Document document, RpcSendQueue q) {
        q.getAndSend(document, d -> d.getSourcePath().toString());
        q.getAndSend(document, d -> d.getCharset().name());
        q.getAndSend(document, Xml.Document::isCharsetBomMarked);
        q.getAndSend(document, Xml.Document::getChecksum);
        q.getAndSend(document, Xml.Document::getFileAttributes);
        q.getAndSend(document, Xml.Document::getProlog, p -> visit(p, q));
        q.getAndSend(document, Xml.Document::getRoot, t -> visit(t, q));
        q.getAndSend(document, Xml.Document::getEof);
        return document;
    }

    @Override
    public Xml visitProlog(Xml.Prolog prolog, RpcSendQueue q) {
        q.getAndSend(prolog, Xml.Prolog::getXmlDecl, d -> visit(d, q));
        q.getAndSendList(prolog, Xml.Prolog::getMisc, m -> m.getId(),
                m -> visit(m, q));
        q.getAndSendList(prolog, Xml.Prolog::getJspDirectives, j -> j.getId(),
                j -> visit(j, q));
        return prolog;
    }

    @Override
    public Xml visitXmlDecl(Xml.XmlDecl xmlDecl, RpcSendQueue q) {
        q.getAndSend(xmlDecl, Xml.XmlDecl::getName);
        q.getAndSendList(xmlDecl, Xml.XmlDecl::getAttributes, a -> a.getId(),
                a -> visit(a, q));
        q.getAndSend(xmlDecl, Xml.XmlDecl::getBeforeTagDelimiterPrefix);
        return xmlDecl;
    }

    @Override
    public Xml visitProcessingInstruction(Xml.ProcessingInstruction pi, RpcSendQueue q) {
        q.getAndSend(pi, Xml.ProcessingInstruction::getName);
        q.getAndSend(pi, Xml.ProcessingInstruction::getProcessingInstructions, c -> visit(c, q));
        q.getAndSend(pi, Xml.ProcessingInstruction::getBeforeTagDelimiterPrefix);
        return pi;
    }

    @SuppressWarnings("unchecked")
    @Override
    public Xml visitTag(Xml.Tag tag, RpcSendQueue q) {
        q.getAndSend(tag, Xml.Tag::getName);
        q.getAndSendList(tag, Xml.Tag::getAttributes, a -> a.getId(),
                a -> visit(a, q));
        q.getAndSendList(tag, t -> (List<Content>) t.getContent(), c -> c.getId(),
                c -> visit(c, q));
        q.getAndSend(tag, Xml.Tag::getClosing, c -> visit(c, q));
        q.getAndSend(tag, Xml.Tag::getBeforeTagDelimiterPrefix);
        return tag;
    }

    @Override
    public Xml visitTagClosing(Xml.Tag.Closing closing, RpcSendQueue q) {
        q.getAndSend(closing, Xml.Tag.Closing::getName);
        q.getAndSend(closing, Xml.Tag.Closing::getBeforeTagDelimiterPrefix);
        return closing;
    }

    @Override
    public Xml visitAttribute(Xml.Attribute attribute, RpcSendQueue q) {
        q.getAndSend(attribute, Xml.Attribute::getKey, k -> visit(k, q));
        q.getAndSend(attribute, Xml.Attribute::getBeforeEquals);
        q.getAndSend(attribute, Xml.Attribute::getValue, v -> visit(v, q));
        return attribute;
    }

    @Override
    public Xml visitAttributeValue(Xml.Attribute.Value value, RpcSendQueue q) {
        q.getAndSend(value, v -> v.getQuote().name());
        q.getAndSend(value, Xml.Attribute.Value::getValue);
        return value;
    }

    @Override
    public Xml visitCharData(Xml.CharData charData, RpcSendQueue q) {
        q.getAndSend(charData, Xml.CharData::isCdata);
        q.getAndSend(charData, Xml.CharData::getText);
        q.getAndSend(charData, Xml.CharData::getAfterText);
        return charData;
    }

    @Override
    public Xml visitComment(Xml.Comment comment, RpcSendQueue q) {
        q.getAndSend(comment, Xml.Comment::getText);
        return comment;
    }

    @Override
    public Xml visitDocTypeDecl(Xml.DocTypeDecl docTypeDecl, RpcSendQueue q) {
        q.getAndSend(docTypeDecl, Xml.DocTypeDecl::getName, n -> visit(n, q));
        q.getAndSend(docTypeDecl, Xml.DocTypeDecl::getDocumentDeclaration);
        q.getAndSend(docTypeDecl, Xml.DocTypeDecl::getExternalId, e -> visit(e, q));
        q.getAndSendList(docTypeDecl, Xml.DocTypeDecl::getInternalSubset, i -> i.getId(),
                i -> visit(i, q));
        q.getAndSend(docTypeDecl, Xml.DocTypeDecl::getExternalSubsets, e -> visit(e, q));
        q.getAndSend(docTypeDecl, Xml.DocTypeDecl::getBeforeTagDelimiterPrefix);
        return docTypeDecl;
    }

    @Override
    public Xml visitDocTypeDeclExternalSubsets(Xml.DocTypeDecl.ExternalSubsets externalSubsets, RpcSendQueue q) {
        q.getAndSendList(externalSubsets, Xml.DocTypeDecl.ExternalSubsets::getElements, e -> e.getId(),
                e -> visit(e, q));
        return externalSubsets;
    }

    @Override
    public Xml visitElement(Xml.Element element, RpcSendQueue q) {
        q.getAndSendList(element, Xml.Element::getSubset, i -> i.getId(),
                i -> visit(i, q));
        q.getAndSend(element, Xml.Element::getBeforeTagDelimiterPrefix);
        return element;
    }

    @Override
    public Xml visitIdent(Xml.Ident ident, RpcSendQueue q) {
        q.getAndSend(ident, Xml.Ident::getName);
        return ident;
    }

    @Override
    public Xml visitJspDirective(Xml.JspDirective jspDirective, RpcSendQueue q) {
        q.getAndSend(jspDirective, Xml.JspDirective::getBeforeTypePrefix);
        q.getAndSend(jspDirective, Xml.JspDirective::getType);
        q.getAndSendList(jspDirective, Xml.JspDirective::getAttributes, a -> a.getId(),
                a -> visit(a, q));
        q.getAndSend(jspDirective, Xml.JspDirective::getBeforeDirectiveEndPrefix);
        return jspDirective;
    }

    @Override
    public Xml visitJspScriptlet(Xml.JspScriptlet jspScriptlet, RpcSendQueue q) {
        q.getAndSend(jspScriptlet, Xml.JspScriptlet::getContent);
        return jspScriptlet;
    }

    @Override
    public Xml visitJspExpression(Xml.JspExpression jspExpression, RpcSendQueue q) {
        q.getAndSend(jspExpression, Xml.JspExpression::getContent);
        return jspExpression;
    }

    @Override
    public Xml visitJspDeclaration(Xml.JspDeclaration jspDeclaration, RpcSendQueue q) {
        q.getAndSend(jspDeclaration, Xml.JspDeclaration::getContent);
        return jspDeclaration;
    }

    @Override
    public Xml visitJspComment(Xml.JspComment jspComment, RpcSendQueue q) {
        q.getAndSend(jspComment, Xml.JspComment::getContent);
        return jspComment;
    }
}

File: rewrite-csharp/csharp/OpenRewrite/Xml/Rpc/XmlSender.cs
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using OpenRewrite.Core;
using OpenRewrite.Core.Rpc;

namespace OpenRewrite.Xml.Rpc;

public class XmlSender : XmlVisitor<RpcSendQueue>
{
    public override Xml? PreVisit(Xml x, RpcSendQueue q)
    {
        q.GetAndSend(x, n => n.Id);
        q.GetAndSend(x, n => n.Prefix);
        q.GetAndSend(x, n => n.Markers);
        return x;
    }

    public override Xml VisitDocument(Document document, RpcSendQueue q)
    {
        q.GetAndSend(document, d => d.SourcePath);
        q.GetAndSend(document, d => d.CharsetName);
        q.GetAndSend(document, d => d.CharsetBomMarked);
        q.GetAndSend(document, d => d.Checksum);
        q.GetAndSend(document, d => d.FileAttributes);
        q.GetAndSend(document, d => d.Prolog, p => Visit(p, q));
        q.GetAndSend(document, d => d.Root, t => Visit(t, q));
        q.GetAndSend(document, d => d.Eof);
        return document;
    }

    public override Xml VisitProlog(Prolog prolog, RpcSendQueue q)
    {
        q.GetAndSend(prolog, p => p.XmlDecl, d => Visit(d, q));
        q.GetAndSendList(prolog, p => p.MiscList, m => m.Id, m => Visit(m, q));
        q.GetAndSendList(prolog, p => p.JspDirectives, j => j.Id, j => Visit(j, q));
        return prolog;
    }

    public override Xml VisitXmlDecl(XmlDecl xmlDecl, RpcSendQueue q)
    {
        q.GetAndSend(xmlDecl, x => x.Name);
        q.GetAndSendList(xmlDecl, x => x.Attributes, a => a.Id, a => Visit(a, q));
        q.GetAndSend(xmlDecl, x => x.BeforeTagDelimiterPrefix);
        return xmlDecl;
    }

    public override Xml VisitProcessingInstruction(ProcessingInstruction pi, RpcSendQueue q)
    {
        q.GetAndSend(pi, p => p.Name);
        q.GetAndSend(pi, p => p.ProcessingInstructions, c => Visit(c, q));
        q.GetAndSend(pi, p => p.BeforeTagDelimiterPrefix);
        return pi;
    }

    public override Xml VisitTag(Tag tag, RpcSendQueue q)
    {
        q.GetAndSend(tag, t => t.Name);
        q.GetAndSendList(tag, t => t.Attributes, a => a.Id, a => Visit(a, q));
        q.GetAndSendList(tag, t => t.ContentList, c => c.Id, c => Visit(c, q));
        q.GetAndSend(tag, t => t.ClosingTag, c => Visit(c, q));
        q.GetAndSend(tag, t => t.BeforeTagDelimiterPrefix);
        return tag;
    }

    public override Xml VisitTagClosing(Tag.Closing closing, RpcSendQueue q)
    {
        q.GetAndSend(closing, c => c.Name);
        q.GetAndSend(closing, c => c.BeforeTagDelimiterPrefix);
        return closing;
    }

    public override Xml VisitAttribute(Attribute attribute, RpcSendQueue q)
    {
        q.GetAndSend(attribute, a => a.Key, k => Visit(k, q));
        q.GetAndSend(attribute, a => a.BeforeEquals);
        q.GetAndSend(attribute, a => a.Val, v => Visit(v, q));
        return attribute;
    }

    public override Xml VisitAttributeValue(Attribute.Value value, RpcSendQueue q)
    {
        q.GetAndSend(value, v => v.QuoteStyle.ToString());
        q.GetAndSend(value, v => v.Val);
        return value;
    }

    public override Xml VisitCharData(CharData charData, RpcSendQueue q)
    {
        q.GetAndSend(charData, c => c.Cdata);
        q.GetAndSend(charData, c => c.Text);
        q.GetAndSend(charData, c => c.AfterText);
        return charData;
    }

    public override Xml VisitComment(Comment comment, RpcSendQueue q)
    {
        q.GetAndSend(comment, c => c.Text);
        return comment;
    }

    public override Xml VisitDocTypeDecl(DocTypeDecl docTypeDecl, RpcSendQueue q)
    {
        q.GetAndSend(docTypeDecl, d => d.Name, n => Visit(n, q));
        q.GetAndSend(docTypeDecl, d => d.DocumentDeclaration);
        q.GetAndSend(docTypeDecl, d => d.ExternalId, e => Visit(e, q));
        q.GetAndSendList(docTypeDecl, d => d.InternalSubset, i => i.Id, i => Visit(i, q));
        q.GetAndSend(docTypeDecl, d => d.ExternalSubsetsNode, e => Visit(e, q));
        q.GetAndSend(docTypeDecl, d => d.BeforeTagDelimiterPrefix);
        return docTypeDecl;
    }

    public override Xml VisitDocTypeDeclExternalSubsets(DocTypeDecl.ExternalSubsets externalSubsets, RpcSendQueue q)
    {
        q.GetAndSendList(externalSubsets, e => e.Elements, el => el.Id, el => Visit(el, q));
        return externalSubsets;
    }

    public override Xml VisitElement(Element element, RpcSendQueue q)
    {
        q.GetAndSendList(element, e => e.Subset, i => i.Id, i => Visit(i, q));
        q.GetAndSend(element, e => e.BeforeTagDelimiterPrefix);
        return element;
    }

    public override Xml VisitIdent(Ident ident, RpcSendQueue q)
    {
        q.GetAndSend(ident, i => i.Name);
        return ident;
    }

    public override Xml VisitJspDirective(JspDirective jspDirective, RpcSendQueue q)
    {
        q.GetAndSend(jspDirective, j => j.BeforeTypePrefix);
        q.GetAndSend(jspDirective, j => j.Type);
        q.GetAndSendList(jspDirective, j => j.Attributes, a => a.Id, a => Visit(a, q));
        q.GetAndSend(jspDirective, j => j.BeforeDirectiveEndPrefix);
        return jspDirective;
    }

    public override Xml VisitJspScriptlet(JspScriptlet jspScriptlet, RpcSendQueue q)
    {
        q.GetAndSend(jspScriptlet, j => j.JspContent);
        return jspScriptlet;
    }

    public override Xml VisitJspExpression(JspExpression jspExpression, RpcSendQueue q)
    {
        q.GetAndSend(jspExpression, j => j.JspContent);
        return jspExpression;
    }

    public override Xml VisitJspDeclaration(JspDeclaration jspDeclaration, RpcSendQueue q)
    {
        q.GetAndSend(jspDeclaration, j => j.JspContent);
        return jspDeclaration;
    }

    public override Xml VisitJspComment(JspComment jspComment, RpcSendQueue q)
    {
        q.GetAndSend(jspComment, j => j.JspContent);
        return jspComment;
    }
}
